### PR TITLE
Cody Gateway: Switch to unquantized StarCoder for MT

### PR DESCRIPTION
**Pull request by philipp-spiess**
This PR adds a new feature flag that determines the rollout of the unquantized StarCoder setups. c.f. https://sourcegraph.slack.com/archives/C0514L18QHG/p1708130736068009

This does not work yet as the 16b unquantized model string doesn't seem to be setup correctly. 

### `CODY_GATEWAY_FIREWORKS_STARCODER_QUANTIZED_PERCENT=100`

```
curl -vS -X POST http://localhost:9992/v1/completions/fireworks -H 'Authorization: bearer <TOKEN>' -d '{"stream":false,"max_tokens":50, "model": "starcoder-16b", "stop_sequences": ["\n\n"], "prompt": "function bubbleSort() {", "stream":false}' -H 'X-sourcegraph-feature: code_completions'
< X-Cody-Resolved-Model: fireworks/accounts/fireworks/models/starcoder-16b-w8a16
```

```
curl -vS -X POST http://localhost:9992/v1/completions/fireworks -H 'Authorization: bearer <TOKEN>' -d '{"stream":false,"max_tokens":50, "model": "starcoder-7b", "stop_sequences": ["\n\n"], "prompt": "function bubbleSort() {", "stream":false}' -H 'X-sourcegraph-feature: code_completions'
< X-Cody-Resolved-Model: fireworks/accounts/fireworks/models/starcoder-7b-w8a16
```

### `CODY_GATEWAY_FIREWORKS_STARCODER_QUANTIZED_PERCENT=0`

```
curl -vS -X POST http://localhost:9992/v1/completions/fireworks -H 'Authorization: bearer <TOKEN>' -d '{"stream":false,"max_tokens":50, "model": "starcoder-16b", "stop_sequences": ["\n\n"], "prompt": "function bubbleSort() {", "stream":false}' -H 'X-sourcegraph-feature: code_completions'
< X-Cody-Resolved-Model: fireworks/accounts/fireworks/models/starcoder-16b

{"id":"cmpl-2fdbe0a3afbe494e88c095e6","object":"text_completion","created":1708425003,"model":"accounts/fireworks/models/starcoder-16b","choices":[{"index":0,"text":"\n                        var num_items = items.length;\n\n                        for (i = 0; i < num_items; i++) {\n\n                            for (j = i + 1; j < num_items; j++) {\n\n                                if (items","logprobs":null,"finish_reason":"length"}],"usage":{"prompt_tokens":5,"total_tokens":55,"completion_tokens":50}}%        

```

```
curl -vS -X POST http://localhost:9992/v1/completions/fireworks -H 'Authorization: bearer <TOKEN>' -d '{"stream":false,"max_tokens":50, "model": "starcoder-7b", "stop_sequences": ["\n\n"], "prompt": "function bubbleSort() {", "stream":false}' -H 'X-sourcegraph-feature: code_completions'
< X-Cody-Resolved-Model: fireworks/accounts/fireworks/models/starcoder-7b

{"id":"cmpl-b556e8be8a2af2deee5ec795","object":"text_completion","created":1708340934,"model":"accounts/fireworks/models/starcoder-7b","choices":[{"index":0,"text":"\n    for (let i = 0; i < list.length; i++) {\n        for (let j = 0; j < list.length - i - 1; j++) {\n            if (list[j] < list","logprobs":null,"finish_reason":"length"}],"usage":{"prompt_tokens":5,"total_tokens":55,"completion_tokens":50}}%   
```

### CODE REVIEW

**Comment by philipp-spiess** [comment]
Converting to draft so we don't ship this yet (since the 16b model strings don't work yet) but feel free to review this @rafax

File: cmd/cody-gateway/internal/httpapi/completions/fireworks.go [hunk 1]
```diff
@@ -131,22 +131,25 @@ func (f *FireworksHandlerMethods) transformBody(body *fireworksRequest, _ string
...
 		body.Model = pickModelBasedOnTrafficSplit(f.config.StarcoderCommunitySingleTenantPercent, fireworks.Starcoder16bSingleTenant, multiTenantModel)
 	}
+
+	// Resolve to the legacy quantized versions if necessary.
+	// TODO: Remove this as soon as the migration to the unquantized models is complete.
+	if body.Model == fireworks.Starcoder16b {
+		body.Model = pickModelBasedOnTrafficSplit(f.config.StarcoderQuantizedPercent, fireworks.Starcoder16b8bit, fireworks.Starcoder16b)
+	} else if body.Model == fireworks.Starcoder7b {
+		body.Model = pickModelBasedOnTrafficSplit(f.config.StarcoderQuantizedPercent, fireworks.Starcoder7b8bit, fireworks.Starcoder7b)
+	}
```
Comment (rafax) [inline cmd/cody-gateway/internal/httpapi/completions/fireworks.go:152]: I think this has gotten to a point where we should extract the "model picking" logic to it's own function and test it with unit-tests - WDYT @philipp-spiess ?

Comment (philipp-spiess) [reply cmd/cody-gateway/internal/httpapi/completions/fireworks.go]: @rafax Like [this](https://github.com/sourcegraph/sourcegraph/pull/60604/commits/26b07d37f77e2ff5b0522ef1876ca3133d419d11)? 👀

**Comment by philipp-spiess** [comment]
The issue is now fixed, all backends work as expected.

**Review by rafax — APPROVED** [review]

**Pull request merged**